File: src/lib/geo.ts
import type { LatLng, Lang, Location } from "@/types";

// Great-circle distance between two points in kilometres (haversine).
export function haversineKm(a: LatLng, b: LatLng): number {
  const R = 6371;
  const toRad = (deg: number) => (deg * Math.PI) / 180;
  const dLat = toRad(b.lat - a.lat);
  const dLng = toRad(b.lng - a.lng);
  const sinLat = Math.sin(dLat / 2);
  const sinLng = Math.sin(dLng / 2);
  const h =
    sinLat * sinLat +
    Math.cos(toRad(a.lat)) * Math.cos(toRad(b.lat)) * sinLng * sinLng;
  return 2 * R * Math.asin(Math.sqrt(h));
}

// "650 м" / "1.2 км" (en: "650 m" / "1.2 km")
export function formatDistance(km: number, lang: Lang): string {
  const [m, kmUnit] = lang === "en" ? ["m", "km"] : ["м", "км"];
  return km < 1 ? `${Math.round(km * 1000)} ${m}` : `${km.toFixed(1)} ${kmUnit}`;
}

export interface NearbyLocation {
  location: Location;
  distanceKm: number;
}

export function nearestTo(origin: LatLng, locations: Location[], count: number): NearbyLocation[] {
  return locations
    .map((location) => ({ location, distanceKm: haversineKm(origin, location.position) }))
    .sort((a, b) => a.distanceKm - b.distanceKm)
    .slice(0, count);
}

export function directionsUrl(from: LatLng, to: LatLng): string {
  const params = new URLSearchParams({
    api: "1",
    origin: `${from.lat},${from.lng}`,
    destination: `${to.lat},${to.lng}`,
  });
  return `https://www.google.com/maps/dir/?${params}`;
}

File: insights/geo.py
import numpy as np

EARTH_RADIUS_KM = 6371.0

CITY_CENTER = (51.1282, 71.4306)
CITY_BOUNDS = {"south": 50.95, "north": 51.32, "west": 71.15, "east": 71.65}


def haversine_km(lat1, lng1, lat2, lng2):
    lat1, lng1, lat2, lng2 = (np.radians(value) for value in (lat1, lng1, lat2, lng2))
    h = (
        np.sin((lat2 - lat1) / 2) ** 2
        + np.cos(lat1) * np.cos(lat2) * np.sin((lng2 - lng1) / 2) ** 2
    )
    return 2 * EARTH_RADIUS_KM * np.arcsin(np.sqrt(h))


def directions_url(origin, lat, lng):
    return (
        "https://www.google.com/maps/dir/?api=1"
        f"&origin={origin[0]},{origin[1]}&destination={lat},{lng}"
    )
